### PR TITLE
Because Luke loves PDFs

### PR DESCRIPTION
**Pull request by humphd**
This adds a new extension to Brackets to allow opening PDF files using PDFJS.  Here's a screenshot of it running, showing that localization into the viewer also works (I'm loading Brackets using the `fr` locale):

screen shot 2017-06-26 at 1 45 12 pm

There are a number of things still to do:
* I need to figure out the right way to include the built PDFJS and the viewer.  What I have now is a mix of doing `gulp build generic` in the pdfjs repo, and also manually copying the `images/`, and `locale/` stuff.  I don't really want to add another submodule, but I also don't think I can use npm (e.g., https://www.npmjs.com/package/pdfjs-dist doesn't have all the web viewer bits I need).
* The viewer app I'm using needs to get [re-skinned](https://github.com/mozilla/pdf.js/wiki/Setup-PDF.js-in-a-website#from-examples).  We can probably strip out a bunch of the features (e.g., we don't need to be able to load a PDF from the viewer).  We can also play with colours, make buttons look more like what you'd expect in Thimble, etc.  I'll need @flukeout's help with that.  You can see what the full viewer looks like [here](http://mozilla.github.io/pdf.js/web/viewer.html).
* I need to sort out the best way to *not* bundle this in our static service worker cache.  The PDF stuff should only get loaded on demand, and then dynamically cached in the service worker.  Right now I am caching everything in `dist/**/*`.  Basically, I need to introduce some way to include an extension without having get processed by the service worker pre-cache build step.  Here's what it looks like when it triggers a load:
screen shot 2017-06-26 at 1 42 22 pm

Finally, I need to see if there's a better way to accomplish the l10n stuff I'm doing in `viewer.html`.  At the very least, this should get moved to an external script vs. being inlined.  I'm not sure if I can do any better than this method, though, since PDFJS seems to block changes to `locale` in the viewer.

Adding @Pomax, who might find this interesting, and also @yurydelendik in case he has any thoughts/tips on the above.

### CODE REVIEW

**Comment by humphd** [comment]
Also, @flukeout, I'll need to add a PDF icon to the filetree before this lands.

**Comment by humphd** [comment]
I've solved the "optional extension caching" issue in https://github.com/mozilla/brackets/pull/818, and I'll move this work to `src/extensions/extra` to take advantage of it.

**Comment by humphd** [comment]
I fixed what I need in order to take advantage of #818.  As predicted, it means a lot here: we would have been caching an extra 250+ files!

**Comment by flukeout** [comment]
Well, this works! Here's my take on this feature: It's useful mainly to check which PDF you have included in your project, in which case here's my proposal.

* Ditch the zooming features and just scale it using transforms to be as wide as possible.
* Limit the UI to...
  * Printing
  * Flipping between pages
* Indicate how big the PDF file is
* Indicate how many pages the PDF has
* For viewing and other features, we can add a "view in new tab" link that uses the browser's built-in viewer.

Thoughts @humphd? I'll work on an icon now.

**Comment by flukeout** [comment]
@humphd - I've attached the icon.

* [pdf.svg.zip](https://github.com/mozilla/brackets/files/1123188/pdf.svg.zip)

So, it looks like the viewer defaults to automatic zoom, which is good for us. I think we can ditch pretty much the whole UI. 

Does it have a mode for a paginated view instead of showing all of the pages at once? If not, is that easy to add? I sort of like how the mac finder preview gives you that UI to go next / previous page...

image

**Comment by humphd** [comment]
@flukeout I've done a bit more work on this.  I think for our needs, the easiest thing to do is to take the existing viewer, and just strip out things we don't need.  I've started on that, and this is probably ready for you to take over.  Some things to note:

* Add any HTML you need to `src/extensions/extra/PDFView/htmlContent/viewer.html`
* CSS can go in `src/extensions/extra/PDFView/styles/viewer-overrides.css` 
* JS can go in `src/extensions/extra/PDFView/bramble-pdfjs.js`

I've stubbed out the API calls you'll need (file size and page info, page navigation, printing, etc).  If you add UI and hook into those, it should just work.

**Comment by flukeout** [comment]
Hey @humphd - sorry I didn't get to this today, will do it first thing tomorrow. The video viewer stuff is taking longer than I thought.

**Comment by flukeout** [comment]
Ok @humphd - got this running locally. Seems to open PDFs just fine.

Questions

* Should I add some placeholder markup for file size, page count and have you populate it?
* How can we handle light and dark themes. The viewer lives inside an iframe, so it's not aware of the theme

**Comment by flukeout** [comment]
So I think at a minimum to get this thing done

* Let's add the filesize and page count above the pages
* Make it work with dark and light themes

Something like...
image

**Comment by humphd** [comment]
OK, I've done what I think you want.  You can make changes to the files I mentioned above for JS/CSS, but I got it started.

**Comment by flukeout** [comment]
@humphd not sure if this is related to your most recent work (or something weird in my env) but I'm not able to load the PDF files in the viewer...

image

It looks like they are 404ing, I can't view them directly via URL either. Might be related to the symptom I'm seeing in the Font Viewer as well.

**Comment by flukeout** [comment]
Disregard the last comment, seems to work fine now, also double-checked that theme switching works - thanks @humphd

**Comment by humphd** [comment]
@gideonthomas do you want to review this before I land?

File: src/extensions/extra/PDFView/PDFView.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+     * @param {!File} file - The PDF file object to render
+     * @param {!jQuery} container - The container to render the image view in
+     */
+    function PDFView(file, $container) {
+        this.file = file;
+        this.$container = $container;
+
+        var self = this;
+
+        file.stat(function(err, stats) {
```
Comment (gideonthomas) [inline src/extensions/extra/PDFView/PDFView.js]: this should probably be moved to a prototype function?

File: src/extensions/extra/PDFView/PDFView.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+     * @param {!Pane} pane - the pane in which to host the view
+     * @return {jQuery.Promise}
+     */
+    function create(file, pane) {
+        var view = pane.getViewForPath(file.fullPath);
+
+        if (view) {
+            pane.showView(view);
+        } else {
+            view = new PDFView(file, pane.$content);
```
Comment (gideonthomas) [inline src/extensions/extra/PDFView/PDFView.js]: any reason this isn't async? If we create the prototype function I mention above, can't we do:
```js
var deferred = new $.Deferred();

if(view) {
    pane.showView(view);
    return deferred.resolve(view.getFile()).promise();
}

view = new PDFView(file, pane.$content);
view.loadFile(function() {
    deferred.resolve(view.getFile());
});

return deferred.promise();
```

File: src/extensions/extra/PDFView/bramble-pdfjs.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    // Size will be a string of the form "256.67 KB"
+    var fileSize = iframe.dataset.fileSize;
+    // Theme will be one of "dark-theme" or "light-theme"
+    var theme = iframe.dataset.theme;
+
+    function parseQueryString(query) {
+        var parts = query.split("&");
+        var params = {};
+        for (var i = 0, ii = parts.length; i < ii; ++i) {
+        var param = parts[i].split("=");
```
Comment (gideonthomas) [inline src/extensions/extra/PDFView/bramble-pdfjs.js]: indentation

Comment (gideonthomas) [inline src/extensions/extra/PDFView/bramble-pdfjs.js]: also might as well switch to:
```js
return parts.reduce(function(params, part) {
    var param = part.split("=");
    var key = param[0].toLowerCase();
    var value = param.length > 1 ? param[1] : null;
    params[decodeURIComponent(key)] = decodeURIComponent(value);
    return params;
}, {});
```

**Pull request merged**